### DISCUSSION TITLE
Label Binarizer를 이용해 categorical feature를 one hot encoding 형태로 만들기!

### DISCUSSION THREAD

**Discussion by MignonDeveloper** [bcaitech1/p4-dkt-freshtomato#109] [Tip]
사실 엄청 간단합니당! One hot encoder 보다는 Label Binarizer를 통해 진행하면 바로 할 수 있어서 좋은 것 같아요.
label binarizer는 아래와 같이 작동합니당. 라벨만 있으면 알아서 원하는 형태로 만들어줍니다.

``` python
from sklearn import preprocessing
lb = preprocessing.LabelBinarizer()
lb.fit([1, 2, 6, 4, 2])

lb.classes_
>>>array([1, 2, 4, 6])

lb.transform([1, 6])
>>>array([[1, 0, 0, 0],
         [0, 0, 0, 1]])
```

저희가 쓰고 있는 __cate_preprocessing 부분을 아래와 같이 변경해줍니다. 이렇게 하면 categorical의 모든 feature에 대해서 one-hot encoding을 진행하니 class가 가지는 label의 값이 너무 많은 경우에는 주의가 필요합니다. final_cate_cols는 새로운 categorical feature의 이름을 의미하며 one hot encoding을 거치면 feature의 개수가 늘어나기 때문에 새롭게 조정이 필요합니다.
```python
from sklearn.preprocessing import MinMaxScaler, LabelBinarizer

def __cate_preprocessing(self, df, cate_cols, is_train=True):
    final_cate_cols = []
    for col in cate_cols:
        ohe = LabelBinarizer()
        ohe.fit(df[col])
        transformed = ohe.transform(df[col])
        ohe_df = pd.DataFrame(transformed, columns=ohe.classes_)
        ohe_df = ohe_df.add_prefix(col)

        df = pd.concat([df, ohe_df], axis=1)
        final_cate_cols.extend(list(ohe_df.columns))

    return df, final_cate_cols
```

새롭게 만들어진 cate_cols 를 받아옵니다.
```python
df, cate_cols = self.__cate_preprocessing(df, cate_cols, is_train)
```

그리고 cate_embs을 지정해줄 때 .npy에서 클래스의 개수를 셀 필요없이 무조건 3개의 값을 가집니다. (원래는 2가지지만, +1을 해주고 마지막에 padding이 들어가기 때문에 총 3가지가 됩니다. 따라서 아래와 같이 원래 코드를 주석해주고 feature 개수에 맞춰서 3을 append 해주면 됩니다.
```python
self.args.cate_embs = []
for cate in cate_cols:
    # self.args.cate_embs.append(len(np.load(os.path.join(self.args.asset_dir, f"{cate}_classes.npy"))))
    self.args.cate_embs.append(3)
```

이렇게 3가지 부분을 수정해주시면 됩니다.
